### DISCUSSION TITLE
Improved Error Messages

### DISCUSSION THREAD

**Discussion by nathanchase** [nuxt/nuxt#19890] [RFCs]
### Describe the feature

I saw that Astro 2.0's error messages were quite lovely, and thought I'd share and pose the question of whether we could make similar enhancements to Nuxt's error messaging.

![image](https://user-images.githubusercontent.com/5963574/215348473-d794e922-8f8e-4b1f-ac59-54d4bbb5cdea.png)

The two biggest wins I see implemented:
- Direct links to docs related specifically to the type of error
- The ability to "open in editor" directly to the file in your code that initiated the error

Beyond that, just the overall aesthetic presentation of the error is well-executed.

I'd be more than happy to assist in any way possible to improve error messaging - particularly on the visual design / presentational side. I'll do some initial digging into the Nuxt code to see how you currently source errors and where the presentation template resides. Would love to hear thoughts from the core team and community on these and other ways to improve the error messaging!

**Comment by Baroshem** [comment]
Hey,

This would be indeed quite useful and helpful in debugging and understanding where the issue was actually caused. Also where to look for the resources to learn more about this problem.

If you don't mind, I would love to take a look at this feature request and prepare some sort of RFC on how this could be implemented.

What do you think @danielroe ? :)

**Comment by pi0** [comment]
Part of this problem will be solved by new cli infra and stack trace utils. Still RFC would be amazing!

**Comment by Baroshem** [comment]
@pi0 

You mean this https://github.com/nuxt/nuxt/issues/18659?

Regarding RFC. I will take a look at other RFC examples in the repo and will work on one over the weekend if that is ok with @danielroe as well :)

**Comment by pi0** [comment]
Browser stack trace handling is probably different which is daniel working on it.

**Comment by Baroshem** [comment]
I see.

I am assigning myself to this issue and will update you with my RFC proposal in a few days :)

**Comment by Baroshem** [comment]
# Better Error Messages in Nuxt RFC

This Request For Comments document is based on the issue raised [here](https://github.com/nuxt/nuxt/discussions/16023#discussioncomment-1683819).

The general idea is to have a better error message in Nuxt that will show the users where they have made a mistake and where to look for additional information such as documentation.

Below, I have summarized my research on Astro errors and how they could be developed in Nuxt. Let me know what you think about it. Once agreed, I could start the implementation part of phase one and later on phase two :)

This is also my first RFC for Nuxt so if there are areas that I should focus more or think about different solutions, please let me know @pi0 @danielroe @manniL . I am open to any feedback as always :)

Inspiration for this RFC taken from the amazing RFC by @lihbr 💚

Labels are going as follows:
- 🟠 Status: to be discussed
- 🔴 Status: rejected
- 🟢 Status: approved

## Error Messages in Astro

I checked how the error messages are created in the Astro framework. All errors are created by using a similar interface:

```ts
export interface ErrorData {
 code: number;
 title: string;
 message?: string | ((...params: any) => string);
 hint?: string | ((...params: any) => string);
}
```

While the error messages themselves is an object with errors looking like this:

```ts
{
 StaticRedirectNotAvailable: {
  title: '`Astro.redirect` is not available in static mode.',
  code: 3001,
  message:
   "Redirects are only available when using `output: 'server'`. Update your Astro config if you need SSR features.",
  hint: 'See https://docs.astro.build/en/guides/server-side-rendering/#enabling-ssr-in-your-project for more information on how to enable SSR.',
 },
}
```

And then, this error is used in the core framework like the following:

```ts
if (missingDimension) {
  throw new AstroError({
    ...AstroErrorData.MissingImageDimension,
    message: AstroErrorData.MissingImageDimension.message(missingDimension, options.src),
  });
}
```

To detect the location where the problem existed and to output it in the error a default error is enhanced like the following (and this is no longer agnostic)

```ts
function enhanceCSSError(err: any, filename: string) {
 const fileContent = fs.readFileSync(filename).toString();
 const styleTagBeginning = fileContent.indexOf(err.input?.source ?? err.code);

 // PostCSS Syntax Error
 if (err.name === 'CssSyntaxError') {
  const errorLine = positionAt(styleTagBeginning, fileContent).line + (err.line ?? 0);

  // Vite will handle creating the frame for us with proper line numbers, no need to create one

  return new CSSError({
   ...AstroErrorData.CSSSyntaxError,
   message: err.reason,
   location: {
    file: filename,
    line: errorLine,
    column: err.column,
   },
   stack: err.stack,
  });
 }
```

I think we could do something similar in Nuxt and I have explained it below :)

## How can we do it in Nuxt

These error messages in Astro are truly impressive and I really think we could have something similar in Nuxt. Astro is using its own Error Codes (like 3005 or similar) for working with errors that should be implemented in Nuxt if we want to support it.

The current [Nuxt error object](https://github.com/nuxt/nuxt/blob/340725bb0c36248e540d35f7b72fe7353854d4d0/packages/nuxt/src/core/runtime/nitro/error.ts#L13-L22) looks like following:

```ts
  const errorObject = {
    url: event.node.req.url,
    statusCode,
    statusMessage,
    message,
    stack: process.dev && statusCode !== 404
      ? `<pre>${stack.map(i => `<span class="stack${i.internal ? ' internal' : ''}">${i.text}</span>`).join('\n')}</pre>`
      : '',
    data: error.data,
  }
```

In my opinion, what we could do with Nuxt Error Messages to make them more DX-friendly while still not too difficult to maintain is to add the following functionalities to the current error objects in two phases.

### Phase One -> Current Error Object

> 🟠 Status: to be discussed

For the current Error Object, I would recommend adding a new property called `location`. It will store the information where the bug was caused in the code based on the stack and can contain a button to locate it in the code as Astro. This should be relatively simple to implement because it just requires better analyzing the error and the stack and displaying data about it in the current error component.

Below is a structure of the error stack:

```bash
ReferenceError: P is not defined                                                                                                                                                                   
    at setup (/Users/jakubandrzejewski/Coding/open-source/nuxt/nuxt-app/app.vue:8:1)
    // stack trace
  {
  statusCode: 500,
  fatal: false,
  unhandled: false,
  statusMessage: undefined,
  __nuxt_error: true
}
```

It contains information about the problem that occurred in while rendering the application. We could analyze this error object in `getErrorLocation` helper function that would extract the data:

```ts
type ErrorLocation = {
  file: string;
  line: number;
  column: string;
}

function getErrorLocation(errorStack: string) => ErrorLocation;
```

It can be used to populate the `location` property in the `errorObject`:

```ts
  const errorObject = {
    url: event.node.req.url,
    statusCode,
    statusMessage,
    message,
    stack: process.dev && statusCode !== 404
      ? `<pre>${stack.map(i => `<span class="stack${i.internal ? ' internal' : ''}">${i.text}</span>`).join('\n')}</pre>`
      : '',
    data: error.data,
    location?: getErrorLocation(stack), // { file, line, column }
  }
```

This data will then serve as the foundation for the following functionalities that need to be approved by the Core Team:

#### Display the file content

> 🟠 Status: to be discussed

Displaying the file content in the browser indicating where the error occurred (it can even be a simple arrow pointing to the place in the code that caused the problem)

From my experience, the first functionality would be more useful as it would show in the browser where the problem occurred. It would be especially useful when developing an application without TypeScript and trying to access a variable that was not registered:

```vue
<template>
  <div>
    <NuxtWelcome />
  </div>
</template>

<script setup>
unregisteredVariable // <- TS would show an error here but for JS apps this will not break until we start the project and inspect the browser
</script>
```

To not make it very advanced in terms of rewriting the current Error component, I have been thinking about basically replacing the currently displayed stack trace with just the name of the file, problem, and the file content with an arrow pointing to the place where the problem occurred.

But I am open to suggestions here :)

#### Display the `Open in Editor` button

> 🟠 Status: to be discussed

It will open the file that caused the problem in the IDE. From my experience this might be an unnecessary feature as for me personally, just having the first option (the content of the file and the location where the problem is) would be enough to allow me to locate the bug and fix it.

## Phase Two -> Enhanced Error Object

> 🟠 Status: to be discussed

As a part of this phase, I would recommend creating three new properties in the `errorObject`:

- `Hint` -> The better definition of the error will be based on the map of errors similar to Astro
- `Internal Code` -> That will allow us to not only operate on 500 errors but more distinguished errors like 3005 from Astro
- `Link to the docs` -> Where to find help about this bug

This would require much more work as to properly implement a set/map of errors we would need to develop internal error codes for all popular issues and adequate hints and links to the docs to fix them.

The structure of the `errorObject` could look like this then:

```ts
  const errorObject = {
    url: event.node.req.url,
    statusCode,
    statusMessage,
    message,
    stack: process.dev && statusCode !== 404
      ? `<pre>${stack.map(i => `<span class="stack${i.internal ? ' internal' : ''}">${i.text}</span>`).join('\n')}</pre>`
      : '',
    data: error.data,
    internalCode: NuxtErrors[error].internalCode, // 3005
    docsLink: NuxtErrors[error].docsLink, // https://nuxt.com/docs/something
    hint: NuxtErrors[error].hint // A component X needs a property `src` to work correctly
  }
```

The `NuxtErrors` object could look like the following:

```ts
const NuxtErrors = {
  <ERROR_NAME>: {
    internalCode: 3005,
    docsLink: 'https://nuxt.com/docs/something',
    hint: 'Component X needs a property `src` to work correctly
  }
  ...
}
```

*I Intentionally haven't created a list of errors yet as I wanted to work on this later on once we will all agree that this approach is correct.*

And then displayed correspondingly in the `error-dev` or `error-500` components. It does not have to be rocket science. A simple paragraph with a hint and a link would do I think:

```
Component X needs the property `src` to work correctly. Learn more [here](https://nuxt.com/docs/something)
```

## Summary

In general, I think and functionality from this RFC would be an improvement to the Developer Experience of using Nuxt. Implementing all of them would elevate error handling to another level in my opinion.

I will appreciate any feedback regarding this feature and the functionality that I have proposed. This is my first RFC so I am sure that it is not perfect. Let me know about your recommendations for the document as well :)

## Resources

- <https://astro.build/blog/astro-2/>
- <https://github.com/withastro/astro/blob/main/packages/astro/src/core/errors/errors-data.ts>
- <https://github.dev/withastro/astro/blob/58c612b746df4b0f489233c71e2aed3f2929acc8/packages/astro/src/core/compile/style.ts#L47-L67>

**Comment by nathanchase** [comment]
@Baroshem Fantastic start! Thank you for taking the time to research and present this so eloquently!

Some considerations:

- Perhaps this could hook into Nuxt DevTools in some capacity, since there's the functionality of running your VSCode editor within Nuxt DevTools.
- When we provide links to send users to the docs to understand how to resolve common issues, perhaps too we may want to include links to the Discussions on the Nuxt framework repo as an additional source (maybe with a search term already pre-populated), and/or deep links to the Nuxt Discord's appropriate channels?
- deciding what the common issues are might be worth it's own separate discussion - taking in all of the various signals of the GH Issues/Discussions, Discord, and perhaps Twitter/Mastodon informal polling to hear developers' pain points.

I don't have any notes on the implementation strategy you've laid out, and I think structuring it as a two-step enhancement makes a lot of sense.

I'm excited at the prospects of this and would be eager to help review and/or contribute.

Again, excellent initiative here and thank you for efforts!

**Reply by Baroshem** [reply]
Thanks @nathanchase for the feedback!

With my research I am always aiming for something relatively easy and doable by me. 

1. Idea with Nuxt DevTools is great but I just didn't have that much time recently to test it out. But this would definitely be a great addition later on (let's say phase 3 -> error framework in the ecosystem).
2. I have been thinking about linking  discussions, discord and more but main problem here is that these links can get out of date really quickly. So putting them in the error hints might require continous maintenance from the core team which I am not sure makes sense. Docs should be the most stable out of all of them :)
3. Agree but I think that we could keep the discussion here for easier discovery of the common issues. 

If you will have any other ideas let us know :)

**Comment by lihbr** [comment]
Awesome job on your first RFC 🎉

Here is my feedback on the proposal, as well as other ideas.

# On the Proposal

## Phase 1

### Re: Adding Location Information

👍 Sounds good and doable to me.

### Re: Displaying the file content

That could be great, I'm a stack trace person however, so I would throw it away so quickly. I think overall we need to discuss the UI/UX of the error page. From my understanding, we're pretty open to changing it in an _extensive_ way.

### Re: Displaying the `Open in Editor` Button

👍

## Phase 2

I don't think we need to be exhaustive about errors before shipping an updated error framework for Nuxt, so no worries about starting to list all of them. I believe this change can be incrementally adopted by the codebase if required~

# Other Ideas and Thoughts

> **Warning**
>
> I'm conscious I go quite moonshot there, I mainly aim to inspire everyone so maybe we can find clever ideas to improve our error framework.

## Additional Reference: Next.js

Just to bring in other references than Astro, I like the way Next.js authors its error messages:

- They have an `errors` directory on their repository, with 1 markdown file = 1 error message
- Those errors get built as part of the documentation, e.g. https://nextjs.org/docs/messages/invalid-new-link-with-extra-anchor
- When an error occurs (iirc):
  - If it's a "fatal", the markdown error gets rendered on the screen
  - If it's not a fatal, a short message pops up in the console with a link to the hosted, longer, error message

---

We adopted a similar à-la-Next.js strategy at Prismic for our SDKs, using instead a `messages` directory so we can also use it for non-error messages (warnings, info during dev, etc.), e.g. https://github.com/prismicio/prismic-react/blob/master/messages/alt-must-be-an-empty-string.md

We don't build those errors on any website so far, we just link to the GitHub markdown file, e.g. https://prismic.dev/msg/react/HEAD/alt-must-be-an-empty-string

## Error Index

Ultimately, I think we'll want, for SEO and other reason, to build an error index on Nuxt documentation, and possibly module documentation (should we find the right framework for it, Docus doc theme could help automate that also)

If we decide to use error codes, I think it might be hard to come up with categories that don't move over time, so just incrementing them like Rust did could maybe be the way to go(?)

References:
- Astro: https://docs.astro.build/en/reference/error-reference
- Rust: https://doc.rust-lang.org/error_codes/error-index.html

## On Markdown

I think the ability to write messages in markdown files comes with multiple benefits:

- Markdown is more accessible for people to edit than TypeScript/programmatic files. At Prismic, we had non-technical people improve/correct our messages thanks to that.
- (not specific to markdown) I think one error = one file is simpler to maintain and browse than a big file of error messages/errors messages scattered across the codebase.

Markdown comes with some challenges though, personalizing errors messages could be harder/we might have to agree on some templating convention to get variables, like `%MODULE_NAME%` to search and replace (or something smarter as I'm not really convinced myself by the previous example ^^')

Ultimately, an error could look like (with the error code and slug extracted from file name):

`./messages/0001-invalid-module-declaration.md`
```markdown
---
title: "Invalid module declaration within Nuxt config."
deprecated: false # Optional, used to indicate the error is not thrown anymore by the latest Nuxt
                  # Alternatively could be a Nuxt version, indicating when the error was deprecated
---

Lorem ipsum `%MODULE_NAME%` sit amet...
```

## Warnings and More

So far we mainly considered fatal/5XX type of errors. I think it could be great to also support other errors, as well as warnings.

For those, they could be advertised in the console with a link to their online/rendered augmented message (if we go markdown).

Because those aren't fatal, the Nuxt devtool could hook to them and have them referenced under a tab/multiple tabs, also rendering their augmented message (also if we go markdown), no need to even get to the doc! (cc @antfu)

## Allowing Modules to Declare Their Own Errors

I think it'll be awesome if we extend the new error framework to also allow modules to declare error messages and such in a standardized way. Nuxt hooks and Nuxt kit could help here.

## Error UIs

You mentioned updating the error template, I think there can be a lot of potential there for some exciting improvements.

### For known errors

<details>
<summary>UI Draft</summary>

![image](https://user-images.githubusercontent.com/25330882/227311167-ba08934b-22dc-4a9c-977c-92ce65470d5c.png)

</details>

For known errors, I think we already shared the essence of what we'd like in the RFC.

I'd likely add a list of places to get help (GitHub discussion, Discord, etc.)

### For unknown errors

<details>
<summary>UI Draft</summary>

![image](https://user-images.githubusercontent.com/25330882/227311228-385e2dd4-b245-4440-800b-98a016d4e26f.png)

</details>

For unknown errors, I think there's potential to help users beyond the stack trace:

- We could try to search for related issues on GitHub, basically googling earlier for the user when possible.
- We could allow users to create an issue from the error report. Since we have the error context we could prefill the error report with every basic information required (we would just need to make sure this doesn't become a way to spam issues on GitHub everytime someone gets an unknown fatal)
- ?

---

Onward 🚀

**Reply by nathanchase** [reply]
I could see in the future perhaps having integrated GPT/Copilot X generated assistance that is trained on the Nuxt docs and this error framework that could present an automated, human-friendly explanation of errors. 

See https://githubnext.com/projects/copilot-for-docs

Obviously at the moment there is an associated cost for the text prompt submissions/responses, but the developer could provide their API key to enable the functionality on demand. At the very least, keeping it as a consideration for a future enhancement seems to have a lot of potential.

**Reply by piqjen-1xofFi-sikxyh** [reply]
Great idea

**Reply by Baroshem** [reply]
Hey Lucie, thank you so much for very detailed comment! 💚

I really like that you have looked at this topic from a different perspective as well. Next.js seem to have a different yet still valid approach. I am up for this idea as well if other core team members will be interested in it too.

For the Error Index, I think the best person to ask about this will be @harlan-zw :)

On Markdown, I think that Nuxt core team an contributors know how to write code so the use case of Markdown might not be perfect here, but it is still probably easier to write hint content in .md instead of JS. It is more a matter of preference basically and I like both ideas.

Warnings and More -> agree but probably in the next phases IMO

Allowing Modules to Declare Their Own Errors -> agree but probably in the next phases IMO

Error UIs -> I have only worked on the current nuxt error design to do some research for my RFC. I am not sure what is the design plan for Nuxt to have for the new components so I am open for any ideas from your side.

**Reply by lihbr** [reply]
Hey Jakub,

I agree Markdown might be a preference on my end, no issue on that ☺️

On the other ideas, should we decide to move forward with some/some variations of them, definitely, I think it's simpler to relay them to other phases 💯

I'd say it sounds reasonably safe to start moving forward with phase one. It will also be an opportunity to confirm/decide where we want to head after that. Let's try to confirm that next insider meeting~

**Reply by Baroshem** [reply]
Sounds awesome! I will rise this discussion on the insiders meeting.

Thank you :)

**Comment by lihbr** [comment]
Hi there, I can confirm we're happy with the team to move with *Phase 1* 🎉 

Few notes:
- For now, we prefer the error location in the error object to be available under an "internal" property name (e.g. `_internalErrorLocation` or `_errorLocation`). Reasons for that being that we don't want users to rely on it for now programmatically and we'll later provide it under a public API following @pi0 work on unified errors (signature might change).
- For displaying the file content and "Open in Editor" button let's see how we can display those with the UI/process we have for now. We'll need to investigate how to refactor the error display to be more consistent across every error context type (sometimes Vue is not available, sometimes JavaScript is broken, etc.), that's an endeavor for later.

If anything sounds off to you please voice it ☺️

**Reply by Baroshem** [reply]
Awesome! 

I will start working then on it this week :)

Thank you for the follow up!
